### PR TITLE
Initial support for backend reindexing visitor functionality

### PR DESCRIPTION
**Pull request by vekterli**
@geirst @toregge please review
@jonmv FYI

Introduces the concept of a read-for-write visitor operation which
blocks all mutating operations from starting for a bucket being
visited. This read-for-write mode is used if (and only if) the visitor
library being specified by the client is `"reindexingvisitor"`.

Since read-for-write visitors cannot race with concurrent write
operations, starting such visitors are deferred until no further
mutations are pending.

Also adds a basic `reindexingvisitor` implementation to the content node
which sends all documents as Puts containing a special TaS token
that will let the operation through even if a bucket is locked. This
token is cleared by the distributor before it is passed on to the
content nodes.

Note: this feature is not yet production ready. For now the following
caveats apply:
* Mutating vs non-mutating pending ops to a bucket are not tracked
  separately, so it’s possible to starve a reindexing visitor by
  sending constant pending read load, as read load is not blocked by
  the operation sequencer.
* Ideal state operations towards locked buckets are not blocked, so
  it's possible for e.g. a split to be sent for a bucket that is being
  visited.

### CODE REVIEW

**Review by toregge — APPROVED** [review]

File: storage/src/vespa/storage/distributor/pendingmessagetracker.h [hunk 1]
```diff
@@ -15,22 +15,48 @@
...
 
-class PendingMessageTracker : public framework::HtmlStatusReporter
-{
+enum class TaskRunState {
+    OK,
+    Aborted,
+    BucketLost
+};
+
+struct DeferredTask {
```
Comment (geirst) [inline storage/src/vespa/storage/distributor/pendingmessagetracker.h:31]: Consider adding class comment.

File: storage/src/vespa/storage/distributor/pendingmessagetracker.h [hunk 2]
```diff
@@ -15,22 +15,48 @@
...
 #include <set>
 #include <unordered_map>
 #include <chrono>
 #include <mutex>
 
 namespace storage::distributor {
 
-class PendingMessageTracker : public framework::HtmlStatusReporter
-{
+enum class TaskRunState {
```
Comment (geirst) [inline storage/src/vespa/storage/distributor/pendingmessagetracker.h:25]: Consider adding class comment.

File: storage/src/vespa/storage/visiting/reindexing_visitor.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#pragma once
+
+#include "visitor.h"
+
+namespace storage {
+
+class ReindexingVisitor : public Visitor {
```
Comment (geirst) [inline storage/src/vespa/storage/visiting/reindexing_visitor.h:8]: Consider adding class comment.

**Review by geirst — APPROVED** [review]
👍

**Comment by vekterli** [comment]
@geirst PTAL at the two latest commits

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**